binary_search_tree/tree.py

```diff
@@ -14,42 +14,164 @@ class Tree:
-        pass
+            if key == current_root.key:
+                return current_root.value
 
-    # Time Complexity: 
-    # Space Complexity: 
+            if key > current_root.key:
+                current_root = current_root.right
+                result = self.find_helper(current_root, key)
+            
+            elif key < current_root.key:
+                current_root = current_root.left
+                result = self.find_helper(current_root, key)
+            return result
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
+    def find(self, key): 
+        if not self.root:
+            return None
+
+        result = self.find_helper(self.root, key)
+
+        return result
+
+
+    def inorder_helper(self, current, items):
+        if current is not None:
+            self.inorder_helper(current.left, items)
+            items.append({"key": current.key, "value": current.value})
+            self.inorder_helper(current.right, items)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+        self.inorder_helper(self.root, items)
+
+        return items
+    
+    def preorder_helper(self, current, items):
+        if current is not None:
+            items.append({"key": current.key, "value": current.value})
+
+            if current.left is not None:
+                self.preorder_helper(current.left, items)
+
+            if current.right is not None:
+                self.preorder_helper(current.right, items)
+        
+        
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper(self.root, items)
+        return items
+
+    def postorder_helper(self, current, items):
+        if current is not None:
+            self.postorder_helper(current.left, items)
+            self.postorder_helper(current.right, items)
+            items.append({"key": current.key, "value": current.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def postorder(self):
-        pass
+        items = []
+        self.postorder_helper(self.root, items)
+        print(items)
+        return items
+
+    def height_helper(self, current, height):
+
+        if current:
 
-    # Time Complexity: 
-    # Space Complexity:     
+            if not current.left and not current.right:
+                return 1
+            
+            else:
+                left = self.height_helper(current.left, height) + 1
+                right = self.height_helper(current.right, height) + 1
+
+                if left > right:
+                    height += left
+                else:
+                    height += right
+        
+        return height
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def height(self):
```

Comment:
✨Nice! However, time and space complexity would be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!